compiler/src/iree/compiler/ConstEval/JitGlobals.cpp

```diff
@@ -146,123 +327,196 @@ struct JitGlobalsPass : public JitGlobalsBase<JitGlobalsPass> {
     options->highLevelOptimizationOptions.constEval = false;
 
     buildIREEVMTransformPassPipeline(
-        // TODO: If ever not using VMVX, plumb the real target registry
-        // through.
-        IREE::HAL::TargetBackendRegistry::getGlobal(), options->bindingOptions,
-        options->inputOptions, options->preprocessingOptions,
-        options->highLevelOptimizationOptions, options->schedulingOptions,
-        options->executableOptions, options->targetOptions, options->hooks,
-        compilePipeline);
+        targetRegistry, options->bindingOptions, options->inputOptions,
+        options->preprocessingOptions, options->highLevelOptimizationOptions,
+        options->schedulingOptions, options->executableOptions,
+        options->targetOptions, options->hooks, compilePipeline);
   }
 
   void getDependentDialects(DialectRegistry &registry) const override {
     compilePipeline.getDependentDialects(registry);
   }
 
+  const SupportedFeatures getSupportedFeatures(MLIRContext *context) {
+    SupportedFeatures s;
+    Builder b(context);
+    s.addScalarType(b.getIntegerType(8));
+    s.addScalarType(b.getIntegerType(16));
+    s.addScalarType(b.getIntegerType(32));
+    s.addScalarType(b.getIntegerType(64));
+    s.addScalarType(b.getF32Type());
+
+    s.addElementType(b.getIntegerType(1));
+    s.addElementType(b.getIntegerType(8));
+    s.addElementType(b.getIntegerType(16));
+    s.addElementType(b.getIntegerType(32));
+    s.addElementType(b.getIntegerType(64));
+    s.addElementType(b.getF32Type());
+    if (!clUseVMVX && hasLLVMCPUBackend) {
+      // The full compilers support additional types.
+      // TODO: Enable support for i4 once it is worked out how to
```

Comment:
this is a good q - it's unfortunate MLIR doesn't store the raw data in the native formats, as our options are either to reencode everything (what we do in IREE::Util::SerializableAttrInterface for taking ElementsAttr -> VMFB) which means reallocating or to insert IR that does the MLIR physical storage -> required logical storage (i4 becomes i8 + trunci to i4 on input). Inserting the IR when prior to fusion and then passing in the raw storage should let us do things zero-copy. If it's inserted after fusion (and we don't somehow perform additional fusion) it'd always require transient memory. Same on the way back out: the execution will return it in properly encoded form but we'll have to get it back into MLIR extended form unless we insert and fuse the IR to do i4 to zexti i8.

Reply:
Yeah, same thought. There may be some other options but I need to talk to Jacques about resources...